### PR TITLE
Added LDAP-search functionality

### PR DESCRIPTION
**Pull request by hermabe**
On the course registration page, the full name field autofills when username is entered.
Also added a function for finding user info from email.

### CODE REVIEW

File: make_queue/templates/make_queue/course/registration_edit.html [hunk 1]
```diff
@@ -81,9 +81,18 @@ <h1>
...
         ;
-        $(".clear-dropdown").click(function() {
+        $(".clear-dropdown").click(function () {
             $(this).closest("div.ui.field").find("div.dropdown").dropdown("clear");
         });
+
+        $("input[name='username']").focusout(function () {
+                $.ajax(langPrefix + "/reservation/json/" + $("input[name='username']").val() + "/", {
+                    success: function (data) {
+                        $("input[name='name']").val(data["full_name"]);
```
Comment (zootos) [inline make_queue/templates/make_queue/course/registration_edit.html]: Could we add a [warning message](https://semantic-ui.com/collections/message.html#warning) when the user cannot be found as well (In similar fashion to the one shown when adding a user)? With text like "The given username could not be found in NTNUs system. Are you sure you entered the correct username?". Would solve many of the problems we currently have with usernames sometimes being entered incorrectly.

File: make_queue/util/user_info.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+from django.contrib.auth.models import User
+import ldap
+
+LDAP_HOST = 'ldap://at.ntnu.no'
```
Comment (zootos) [inline make_queue/util/user_info.py]: Could we move this file to the `dataporten` app with a file name of `ldap.py`? It feels more natural to collect all the external user information sources in one app. Additionally, adding [docstrings](https://www.python.org/dev/peps/pep-0257/) would be great.

File: make_queue/util/user_info.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+def get_user_details_from_email(email):
+    user = User.objects.filter(email=email).first()
+    if user:
+        return get_user_details_from_username(username=user.username)
+
+    ldap_results = LDAP_search("mail", email)
+    if len(ldap_results) == 0:
+        return None
+    return get_user_details_from_username(get_LDAP_field(ldap_results, "username"))
```
Comment (zootos) [inline make_queue/util/user_info.py]: This could result in two searches to the database, while the information should be available after the first search. Would be better to move the retrieval of data from the LDAP results to a helper function and instead call that.

File: make_queue/util/user_info.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    return user_details
+
+
+def get_user_details_from_email(email):
+    user = User.objects.filter(email=email).first()
+    if user:
+        return get_user_details_from_username(username=user.username)
+
+    ldap_results = LDAP_search("mail", email)
+    if len(ldap_results) == 0:
```
Comment (zootos) [inline make_queue/util/user_info.py]: Invert to
```python3
if ldap_results:
    return get_user_details_from_username(get_LDAP_field(ldap_results, "username"))
return {}
```
as that is more readable.

File: make_queue/util/user_info.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        if len(ldap_results) == 0:
+            return None
+        user_details["full_name"] = get_LDAP_field(ldap_results, "full_name")
+        user_details["email"] = get_LDAP_field(ldap_results, "mail")
+
+    return user_details
+
+
+def get_user_details_from_email(email):
+    user = User.objects.filter(email=email).first()
```
Comment (zootos) [inline make_queue/util/user_info.py]: Would be great if one could specify to retrieve the data from the LDAP server even though the user exists locally. E.g. if I want to find the NTNU username and not the derived username we get through the email supplied from dataporten. Could probably just use a kwarg to indicate if we want to use cached (local data). 

```python3
def get_user_details_from_email(email, use_cached=True):
```

A similar feature would also be great for the username search.

File: make_queue/util/user_info.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+def LDAP_search(field, value):
+    l = ldap.initialize(LDAP_HOST, bytes_mode=False)
+    l.simple_bind()
+    query = "({}={})".format(LDAP_FIELDS[field], value)
+    return l.search_s(LDAP_BASE, ldap.SCOPE_SUBTREE, query)
+
+
+def get_LDAP_field(ldapData, field):
+    return ldapData[0][1].get(LDAP_FIELDS[field], [b''])[0].decode()
+    # list<tuple<str, dict<str, list<bstr>>>>
```
Comment (zootos) [inline make_queue/util/user_info.py]: I think it would be more useful with a description of the data structure (how is everything packed) we receive from LDAP, rather than just the datatypes. E.g. "We get a list of results. These consists of tuples of...".

File: make_queue/util/user_info.py (hunk 5 above)
Comment (hermabe) [reply make_queue/util/user_info.py]: Have added the option to only use LDAP search. Can also use the `get_user_details_from_LDAP(search_field, search_value)` function directly.

File: make_queue/util/user_info.py (hunk 2 above)
Comment (hermabe) [reply make_queue/util/user_info.py]: Moved to `dataporten/ldap_utils.py` to not shadow the `python-ldap` package.

File: make_queue/templates/make_queue/course/registration_edit.html (hunk 1 above)
Comment (hermabe) [reply make_queue/templates/make_queue/course/registration_edit.html]: Added a large error message at the top of the page when the user is not found. The users should also be able to verify that the username entered matches with the full name field, so hope this becomes lesser of a problem in the future.

File: make_queue/templates/make_queue/course/registration_edit.html [hunk 7]
```diff
@@ -25,6 +25,15 @@ <h1>
         </h1>
 
         {% block messages %}
+            <div id="messages" style="margin-bottom:20px; display: none;">
+                <div class="ui red message">
+                    <i class="close icon"></i>
```
Comment (zootos) [inline make_queue/templates/make_queue/course/registration_edit.html]: I think the outer div is not needed. If you remove it, the `message` class already has a margin on the bottom. And you can use `ui hidden red message` as the class, to hide the message initially. Then just us an id on the `ui red message` div to easily select it with jQuery.

Might also be better to use `warning` instead of `red`, as this is not necessarily an error (Dataporten is not limited to NTNU users only, however, LDAP is).

File: make_queue/templates/make_queue/course/registration_edit.html [hunk 8]
```diff
@@ -73,17 +82,24 @@ <h1>
...
-            .on('click', function () {
-                $(this)
-                    .closest('.message')
-                    .transition('fade')
-                ;
-            })
-        ;
-        $(".clear-dropdown").click(function() {
+        $('#messages').click(function () {
+            $(this).fadeOut();
```
Comment (zootos) [inline make_queue/templates/make_queue/course/registration_edit.html]: This change away from `$('.message .close')...` means that the message for successful adds will not be closable. So you could probably just revert this change and everything (with the adjustments from the other comment) should work just fine.

Comment (hermabe) [reply make_queue/templates/make_queue/course/registration_edit.html]: I completely forot about the confirmation message. Have fixed it now.

File: dataporten/ldap_utils.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        """
+    if use_cached:
+        user = User.objects.filter(email=email).first()
+        if user:
+            return {
+                "username": user.username,
+                "full_name": user.get_full_name(),
+                "email": email,
+            }
+    return get_user_details_from_LDAP("mail", email)
```
Comment (zootos) [inline dataporten/ldap_utils.py]: This errors. Should probably be `"email"`.

Comment (hermabe) [reply dataporten/ldap_utils.py]: Yes, forgot to change that.
